server/dash/dashboards/nc_summary_comparison.py

```diff
@@ -0,0 +1,754 @@
+    if not nc_dropdown:
+        df = api_data_df
+    else:
+        df = api_data_df[api_data_df["councilName"] == nc_dropdown]
+    # Filter as per selection on Request Type Dropdown.
+    if not nc_dropdown_filter and not nc_dropdown:
+        df = api_data_df
+    elif nc_dropdown_filter:
+        df = df[df["typeName"].isin(nc_dropdown_filter)]
+    return df
+
+
+def filter_bad_quality_data(df, data_quality_switch=True):
+    """Filters the dataframe based on pre-defined data quality filters.
+
+    This functions takes the original dataframe "df" and filters out records 
+    with an outlier amount 
+    of request time to close based on the Freedman-Diaconis Rule.
+
+    Args:
+        df: 311-request data accessed from the API.
+        data_quality_switch: A string argument automatically detected by Dash 
+            callback function when "data_quality_switch" toggle element is selected in layout.
+        nc_dropdown_filter: A list of strings automatically detected by Dash 
+            callback function when "nc_dropdown_filter" element is selected in the 
+            layout, default None.
+    Returns:
+        df: filtered dataframe excluding outliers.
+        num_bins: the number of bins that will be used to plot histogram.
+        data_quality_output: a string being displayed on whether the data quality switch is on or off.
+    """
+    print("* Getting quality data.")
+    df.loc[:, "createDateDT"] = pd.to_datetime(
+        df.loc[:, "createdDate"].str[:-4].str.split("T").str.join(" "))
+    df.loc[:, "closeDateDT"] = pd.to_datetime(
+        df.loc[:, "closedDate"].str[:-4].str.split("T").str.join(" "))
+    df.loc[:, "timeToClose"] = (df.loc[:, "closeDateDT"] - df.loc[:, "createDateDT"]).dt.days
+
+    # Calculate the Optimal number of bins based on Freedman-Diaconis Rule.
+
+    # Replace empty rows with 0.0000001 To avoid log(0) error later.
+    df.loc[:, "timeToClose"] = df.loc[:, "timeToClose"].fillna(0.0000001)
+
+    # Replace negative values
+    df = df[df["timeToClose"] > 0]
+    if df.shape[0] == 0:
+        raise PreventUpdate()
+    else:
+        q3, q1 = np.percentile(df.loc[:, "timeToClose"].astype(int), [75, 25])
+        iqr = q3 - q1
+        if not iqr:
+            num_bins = 100
+        else:
+            num_bins = int((2 * iqr) / (df.shape[0]**(1 / 3)))
+
+        # Log Transform, Compute IQR, then exclude outliers.
+        df.loc[:, "logTimeToClose"] = np.log(df.loc[:, "timeToClose"])
+        log_q3, log_q1 = np.percentile(df.loc[:, "logTimeToClose"], [75, 25])
+        log_iqr = log_q3 - log_q1
+
+    # Data Quality switch to remove outliers as defined by Median +- 1.5*IQR.
+    if data_quality_switch:
+        temp = df[(df.loc[:, "logTimeToClose"] > 1.5 * log_iqr - np.median(
+            df.loc[:, "logTimeToClose"])) &
+            (df.loc[:, "logTimeToClose"] < 1.5 * log_iqr + np.median(
+                df.loc[:, "logTimeToClose"]))]
+        if temp.shape[0] > 0:
```

Comment:
Is it possible that all rows would be filtered out (i.e., `temp.shape[0] == 0`)?

Reply:
Yes, in which case we use the original dataframe.